# Ejercicio Formativo 2 Capítulo 1

In [1]:
import json

In [2]:
with open("movies.json", encoding="utf8") as movies_file:
    movies = json.load(movies_file)

## Analizando movies y su contenido

In [3]:
print(f"movies es del tipo {type(movies)}")
print(f"Tiene {len(movies)} elementos")

movies es del tipo <class 'list'>
Tiene 28795 elementos


In [4]:
print(f"El primer elemento es {movies[0]}")
print(f"El primer elemento es del tipo {type(movies[0])}")

El primer elemento es {'title': 'After Dark in Central Park', 'year': 1900, 'cast': [], 'genres': []}
El primer elemento es del tipo <class 'dict'>


Entonces movies es una lista de diccionarios, donde cada diccionario representa una película. Cada película tiene un título, un año de lanzamiento, una lista de actores (cast) y una lista de géneros (genres).

Entonces, para acceder a la información de una película, primero accedemos a la lista de películas y luego al índice de la película que queremos ver. Por ejemplo, para acceder al título de la primera película, usamos `movies[0]['title']` y se haría lo mismo para acceder a los otros datos

## Creando Clases

In [5]:
class Actor:
    def __init__(self, full_name, start_career):
        self.full_name = full_name
        self.n_movies = 0
        self.start_career = start_career
        self.end_career = start_career

    def update_track_record(self, year):
        if self.start_career > year:
            self.start_career = year
        if self.end_career < year:
            self.end_career = year
            
    def years_track_record(self):
        return self.end_career - self.start_career

    def __repr__(self):
        return f"{self.full_name}"
    
class Genre:
    def __init__(self, name):
        self.name = name
        self.n_movies = 0
    
    def __repr__(self):
        return f"{self.name}"

class Movie:
    def __init__(self, title, year, cast, genres):
        self.title = title
        self.year = year
        self.cast = cast
        self.genres = genres
        self.add_info()

    def add_info(self):
        for actor in self.cast:
            actor.n_movies += 1
        for genre in self.genres:
            genre.n_movies += 1
    
    def __repr__(self):
        return f"{self.title} ({self.year}) - {self.cast} - {self.genres}"

## Cargando datos a objetos

Se decidió que tanto para los actores, los géneros y las películas se utilizarían diccionarios para guardar los objetos de cada tipo.

In [6]:
actors_dict = {}
genres_dict = {}
movies_dict = {}

Se recorre la lista de películas y se crea un objeto de las clases correspondientes para cada actor, género y película. Luego se agrega cada uno de ellos a su respectivo diccionario.

In [7]:
for movie in movies:
    cast = []
    genres = []
    for actor in movie["cast"]:
        if actor not in actors_dict:
            actors_dict[actor] = Actor(actor, movie["year"])
        else:
            actors_dict[actor].update_track_record(movie["year"])
        cast.append(actors_dict[actor])
    for genre in movie["genres"]:
        if genre not in genres_dict:
            genres_dict[genre] = Genre(genre)
        genres.append(genres_dict[genre])
    if (movie["title"], movie["year"],tuple(cast)) not in movies_dict:
        movies_dict[(movie["title"], movie["year"], tuple(cast))] = Movie(movie["title"], movie["year"], tuple(cast), genres)
    else:
        print("*"*80)
        print(f"Película duplicada: {movie['title']} - ({movie['year']}) - {tuple(cast)} - {genres}")
        movie_found = movies_dict[(movie["title"], movie["year"], tuple(cast))]
        print(f"En el diccionario ya está: {movie_found}")
        print("*"*80)


********************************************************************************
Película duplicada: The Prince and Betty - (1919) - (William Desmond, Mary Thurman) - [Comedy]
En el diccionario ya está: The Prince and Betty (1919) - (William Desmond, Mary Thurman) - [Comedy]
********************************************************************************
********************************************************************************
Película duplicada: The Virtuous Thief - (1919) - (Enid Bennett, Niles Welch) - [Drama]
En el diccionario ya está: The Virtuous Thief (1919) - (Enid Bennett, Niles Welch) - [Drama]
********************************************************************************
********************************************************************************
Película duplicada: The Wise Kid - (1922) - (Gladys Walton, David Butler) - [Comedy]
En el diccionario ya está: The Wise Kid (1922) - (Gladys Walton, David Butler) - [Comedy]
********************************************

Para almacenar los objetos de las películas se utiliza un diccionario, donde la llave es una tupla que contiene el título de la película, el año de lanzamiento y una tupla con los nombres de los actores. Dicho esto se encontraron películas con el mismo título, año de lanzamiento y actores, para evitar tener dos objetos de la misma película se decidió utilizar un `if` para verificar si la película ya estaba en el diccionario de películas.

In [10]:
print(f"movies_dict tiene {len(movies_dict)} elementos")

movies_dict tiene 28789 elementos


Se puede observar que el diccionario tiene almacenados 28789 elementos, lo cual es el número de películas que se encuentran en `movies`, sin contar las 6 repeticiones. Por ende la creación de objetos de las películas se realizó de manera correcta.

## Consultas

### Encuentre los 5 géneros más populares

In [11]:
def popular_genres(genres_dict):
    sorted_list_genres = sorted(genres_dict.values(), key=lambda x: x.n_movies, reverse=True)
    most_popular_genres = sorted_list_genres[:5]
    print("Los 5 géneros más populares son:")
    for genre in most_popular_genres:
        print(f"{genre.name} hay {genre.n_movies} películas de este género")


In [12]:
popular_genres(genres_dict)

Los 5 géneros más populares son:
Drama hay 8742 películas de este género
Comedy hay 7361 películas de este género
Western hay 3011 películas de este género
Crime hay 1499 películas de este género
Horror hay 1166 películas de este género


### Encuentre los 3 años con más películas estrenadas.

In [13]:
def years_most_premiers(movies_dict):
    premiers_dict = {}
    for movie in movies_dict.values():
        if movie.year not in premiers_dict:
            premiers_dict[movie.year] = 0
        premiers_dict[movie.year] += 1
    sorted_list_years = sorted(premiers_dict.items(), key=lambda x: x[1], reverse=True)
    list_years = sorted_list_years[:3]
    print("Los 3 años con más películas estrenadas son:")
    for data in list_years:
        print(f"El año {data[0]} con {data[1]} películas producidas")

In [14]:
years_most_premiers(movies_dict)

Los 3 años con más películas estrenadas son:
El año 1919 con 632 películas producidas
El año 1925 con 572 películas producidas
El año 1936 con 504 películas producidas


### Encuentre a los 5 actores con la trayectoria más larga, es decir, mayor cantidad de años actuando

In [15]:
def longer_track_record(actors_dict):
    sorted_list_actors = sorted(actors_dict.values(), key=lambda x: x.years_track_record(), reverse=True)
    list_actors = sorted_list_actors[:5]
    print("Los 5 actores con trayectoria más larga son:")
    for actor in list_actors:
        print(f"{actor.full_name} con {actor.years_track_record()} años de trayectoria")

In [16]:
longer_track_record(actors_dict)

Los 5 actores con trayectoria más larga son:
. con 101 años de trayectoria
and con 98 años de trayectoria
Harrison Ford con 98 años de trayectoria
Gloria Stuart con 80 años de trayectoria
Lillian Gish con 75 años de trayectoria


### Encuentre el reparto de una película (2 o más actores) que se haya repetido completo en otras la mayor cantidad de veces.

In [17]:
def most_repeated_casts(movies_dict):
    cast_dict = {}
    for movie in movies_dict.values():
        if len(movie.cast) >= 2:
            if movie.cast not in cast_dict:
                cast_dict[movie.cast] = 0
            cast_dict[movie.cast] += 1
        else:
            continue
    max_value = max(cast_dict.items(), key=lambda x: x[1])
    print(f"El reparto de una película que más se ha repetido es {max_value[0]} con {max_value[1]} repeticiones")

In [18]:
most_repeated_casts(movies_dict)

El reparto de una película que más se ha repetido es (Harold Lloyd, Bebe Daniels) con 44 repeticiones
